### PR TITLE
Apply uniform contractions to starred rows/columns

### LINKED ISSUE #8797: [Bug] Word wrapped Label not measured correctly
### Description
A Label in a grid that is contained in a StackLayout isn't measured correctly, with contents below the grid in the StackLayout overlapping the grid and label.

### Steps to Reproduce

1. Create a new template XF solution.
2. Replace the MainPage xml with a new MainPage.cs
3. Replace the code with the following and fix the namespace accordingly:

```
using System;
using Xamarin.Forms;
using Xamarin.Forms.PlatformConfiguration;
using Xamarin.Forms.PlatformConfiguration.iOSSpecific;

namespace LabelSizeTest
{
    public class MainPage : ContentPage
    {
        public MainPage()
        {
            On<iOS>().SetUseSafeArea(true);

            var grid = new Grid();
            grid.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Star });
            grid.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Star });
            grid.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Star });
            grid.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Star });
            grid.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Star });
            grid.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Star });
            grid.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Star });
            grid.RowDefinitions.Add(new RowDefinition() { Height = GridLength.Auto });

            var label = new Label();
            label.LineBreakMode = LineBreakMode.WordWrap;
            label.Text = "This is a test teest test 2nd 3rd last";

            grid.Children.Add(label, 1, 0);

            var stack = new StackLayout()
            {
                VerticalOptions = LayoutOptions.FillAndExpand,
                HorizontalOptions = LayoutOptions.FillAndExpand,
            };

            stack.Children.Add(new BoxView()
            {
                BackgroundColor = Color.Blue,
                HeightRequest = 50
            });

            stack.Children.Add(grid);

            stack.Children.Add(new BoxView()
            {
                BackgroundColor = Color.Blue,
                HeightRequest = 50
            });

            Content = stack;
        }
    }
}
```
4. Run on the iPhone 11 Simulator.

### Expected Behavior
Label should display correctly, with all the text visible.

### Actual Behavior
Label is being partially obscured by the 2nd BoxView, the words 3rd and Last are replaced with the BoxView.

### Basic Information

- Version with issue:
- Last known good version:
- IDE: VS4Mac
- Platform Target Frameworks: 
  - iOS:  

- Nuget Packages: XF 4.2.0
- Affected Devices:

### Screenshots

Simulator Screen Shot - iPhone 11 - 2019-12-08 at 11 15 31

### PR DESCRIPTION
**Pull request by hartez**
### Description of Change ###

The Grid attempts to contract columns/rows to fit into the targeted layout. When doing this, it uses the same logic for contracting Star columns/rows as it does for Auto columns/rows. However, Star columns/rows are supposed to equally divide the remaining space; i.e., all 1* columns/rows should be the same size. 

The algorithm for contracting Auto columns/rows does so proportionally based on the difference between the current size and the minimum size of the contents. This is fine for Auto columns/rows, but for Stars this causes a problem. If the minimum sizes of the contents of the columns/rows vary, then the amount of contraction applied also varies; this means that the Star columns/rows can end up with different sizes.

In addition to violating the definition of Star columns/rows, this can also cause a nasty problem with multi-line text with wrapping turned on; a Label can be measured multiple times with contracting widths. If the Label is also defining the height of a containing Grid, the second measurement pass and subsequent layout will use the narrower width from the second pass and the height from the first pass. This forces the bottom of the text to be cut off. 

Anyway, these changes fix the problem. The contraction of Starred columns/rows is now done in a separate method than the Auto columns/rows; the new method respects the requirements for equal size.

### Issues Resolved ### 

- fixes #8797

### API Changes ###
 
None

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

Unit tests, plus a manual test (Issue 8797 in Control Gallery)

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: .Xamarin.Forms.Android.slnf [hunk 1]
```diff
@@ -22,14 +19,12 @@
...
       "Xamarin.Forms.Pages.Azure\\Xamarin.Forms.Pages.Azure.csproj",
-      "Xamarin.Forms.Pages.UnitTests\\Xamarin.Forms.Pages.UnitTests.csproj",
       "Xamarin.Forms.Pages\\Xamarin.Forms.Pages.csproj",
       "Xamarin.Forms.Platform.Android.AppLinks\\Xamarin.Forms.Platform.Android.AppLinks.csproj",
       "Xamarin.Forms.Platform.Android.FormsViewGroup\\Xamarin.Forms.Platform.Android.FormsViewGroup.csproj",
+      "Xamarin.Forms.Platform.Android.UnitTests\\Xamarin.Forms.Platform.Android.UnitTests.csproj",
       "Xamarin.Forms.Platform.Android\\Xamarin.Forms.Platform.Android.csproj",
       "Xamarin.Forms.Platform\\Xamarin.Forms.Platform.csproj",
-      "Xamarin.Forms.Sandbox.Android\\Xamarin.Forms.Sandbox.Android.csproj",
-      "Xamarin.Forms.Sandbox\\Xamarin.Forms.Sandbox.csproj",
```
Comment (jsuarezruiz) [inline .Xamarin.Forms.Android.slnf:32]: Are these changes necessary?

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8797.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	[Issue(IssueTracker.Github, 8797, "[Bug] Word wrapped Label not measured correctly",
+		PlatformAffected.Android | PlatformAffected.iOS)]
+	public class Issue8797 : TestContentPage
+	{
+		protected override void Init()
+		{
+			var scrollView = new ScrollView();
+			var layout = new StackLayout { Margin = new Thickness(5, 80, 5, 0) };
+
+			var instructions = new Label { Text = "The text visible in the Grid should end with 'finding a good way to spend it'. If that text is cut off, this test has failed." };
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8797.cs:14]: This is working fine!
Captura de pantalla 2020-07-08 a las 12 58 02

File: .Xamarin.Forms.Android.slnf (hunk 1 above)
Comment (hartez) [reply .Xamarin.Forms.Android.slnf:32]: No, I'll revert that.

Comment (hartez) [reply .Xamarin.Forms.Android.slnf:32]: Oh, wait - yes, they are necessary. The solution filter was still trying to include stuff we removed since it was created.
